source/aws_imds_client.c

```diff
@@ -951,3 +951,664 @@ int aws_imds_client_get_resource_async(
+ */
+static struct aws_byte_cursor s_instance_identity_document =
+    AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("instance-identity/document");
+static struct aws_byte_cursor s_instance_identity_signature =
+    AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("instance-identity/signature");
+static struct aws_byte_cursor s_ec2_metadata_root = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("/latest/meta-data");
+static struct aws_byte_cursor s_ec2_credentials_root =
+    AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("/latest/meta-data/iam/security-credentials/");
+static struct aws_byte_cursor s_ec2_userdata_root = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("/latest/user-data/");
+static struct aws_byte_cursor s_ec2_dynamicdata_root = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("/latest/dynamic/");
+
+struct imds_get_array_user_data {
+    struct aws_allocator *allocator;
+    aws_imds_client_on_get_array_callback_fn *callback;
+    void *user_data;
+};
+
+struct imds_get_credentials_user_data {
+    struct aws_allocator *allocator;
+    aws_imds_client_on_get_credentials_callback_fn *callback;
+    void *user_data;
+};
+
+struct imds_get_iam_user_data {
+    struct aws_allocator *allocator;
+    aws_imds_client_on_get_iam_profile_callback_fn *callback;
+    void *user_data;
+};
+
+struct imds_get_instance_user_data {
+    struct aws_allocator *allocator;
+    aws_imds_client_on_get_instance_info_callback_fn *callback;
+    void *user_data;
+};
+
+static void s_process_array_resource(const struct aws_byte_buf *resource, int error_code, void *user_data) {
+    struct imds_get_array_user_data *wrapped_user_data = user_data;
+    struct aws_array_list resource_array;
+    AWS_ZERO_STRUCT(resource_array);
+
+    if (resource && !error_code) {
+        struct aws_byte_cursor resource_cursor = aws_byte_cursor_from_buf(resource);
+        if (aws_array_list_init_dynamic(
+                &resource_array, wrapped_user_data->allocator, 10, sizeof(struct aws_byte_cursor))) {
+            goto on_finish;
+        }
+        aws_byte_cursor_split_on_char(&resource_cursor, '\n', &resource_array);
+    }
+
+on_finish:
+    wrapped_user_data->callback(&resource_array, error_code, wrapped_user_data->user_data);
+    aws_array_list_clean_up_secure(&resource_array);
+    aws_mem_release(wrapped_user_data->allocator, wrapped_user_data);
+}
+
+static void s_process_credentials_resource(const struct aws_byte_buf *resource, int error_code, void *user_data) {
+    struct imds_get_credentials_user_data *wrapped_user_data = user_data;
+    struct aws_credentials *credentials = NULL;
+
+    struct aws_byte_buf json_data;
+    AWS_ZERO_STRUCT(json_data);
+
+    if (!resource || error_code) {
+        goto on_finish;
+    }
+
+    if (aws_byte_buf_init_copy(&json_data, wrapped_user_data->allocator, resource)) {
+        goto on_finish;
+    }
+
+    if (aws_byte_buf_append_null_terminator(&json_data)) {
+        goto on_finish;
+    }
+
+    struct aws_parse_credentials_from_json_doc_options parse_options = {
```

Comment:
I wonder if there's a generalization here for a parse function that takes the json doc and an array of mappings where a mapping is either a (json path -> pointer to a byte cursor in the target struct) or (json path -> pointer to an array list (of byte cursors) in the target struct) or (json path -> pointer to date time) that would let a single function do the parsing with just different array configurations to handle the different docs/substructures.  Not required just something to consider; you'd lose some of the caching (refollowing the same path repeatedly) unless you added additional complexity.

Also, I imagine this data changes very slowly from a modeling perspective, but it's a little uneasy to have the model hard-coded.  But if you don't provide that modeling then there's nothing to do really.  I'd check with henso to make sure this seems reasonable.  I don't know what other ec2 metadata clients do.

Reply:
Justin mentioned in the stand up that this would be our next year target to do data serialization/deserialization.